src/buffered_range_client.rs

```diff
@@ -54,34 +54,383 @@ impl BufferedHttpRangeClient {
+
+        let bytes = client.get_range(205670, 20).await?;
+        assert_eq!(bytes, [78, 192, 205, 204, 204, 204, 204, 236, 73, 192]);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn custom_headers() -> Result<()> {
+        init_logger();
+        let http_client = reqwest::Client::builder()
+            .user_agent("rust-client")
+            .build()
+            .unwrap();
+        let mut client = AsyncBufferedHttpRangeClient::with(
+            http_client,
+            "https://flatgeobuf.org/test/data/countries.fgb",
+        );
+        let bytes = client.min_req_size(256).get_range(0, 3).await?;
+        assert_eq!(bytes, b"fgb");
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+#[cfg(feature = "reqwest-sync")]
+mod test_sync {
+    use crate::{HttpReader, Result};
+    use std::io::{Read, Seek, SeekFrom};
+
+    fn init_logger() {
+        let _ = env_logger::builder().is_test(true).try_init();
+    }
+
+    #[test]
+    fn http_read_sync() -> Result<()> {
+        init_logger();
+        let mut client = HttpReader::new("https://flatgeobuf.org/test/data/countries.fgb");
+        let bytes = client.min_req_size(256).get_range(0, 3)?;
+        assert_eq!(bytes, b"fgb");
+
+        let version = client.get_bytes(1)?;
+        assert_eq!(version, [3]);
+
+        let bytes = client.get_bytes(3)?;
+        assert_eq!(bytes, b"fgb");
+        Ok(())
+    }
+
+    #[test]
+    fn http_read_sync_zero_range() -> Result<()> {
+        init_logger();
+        let mut client = HttpReader::new("https://flatgeobuf.org/test/data/countries.fgb");
+        let bytes = client.min_req_size(256).get_range(0, 0)?;
+        assert_eq!(bytes, []);
+        Ok(())
+    }
+
+    #[test]
+    fn io_read() -> std::io::Result<()> {
+        init_logger();
+        let mut client = HttpReader::new("https://flatgeobuf.org/test/data/countries.fgb");
+        client.seek(SeekFrom::Start(3)).ok();
+        let mut version = [0; 1];
+        client.min_req_size(256).read_exact(&mut version)?;
+        assert_eq!(version, [3]);
+
+        let mut bytes = [0; 3];
+        client.read_exact(&mut bytes)?;
+        assert_eq!(&bytes, b"fgb");
+        Ok(())
+    }
+
+    #[test]
+    fn io_read_over_min_req_size() -> std::io::Result<()> {
+        init_logger();
+        let mut client = HttpReader::new("https://flatgeobuf.org/test/data/countries.fgb");
+        let mut bytes = [0; 8];
+        client.min_req_size(4).read_exact(&mut bytes)?;
+        assert_eq!(bytes, [b'f', b'g', b'b', 3, b'f', b'g', b'b', 0]);
+        Ok(())
+    }
+
+    #[test]
+    fn io_read_non_exact() -> std::io::Result<()> {
+        init_logger();
+        let mut client = HttpReader::new("https://flatgeobuf.org/test/data/countries.fgb");
+        let mut bytes = [0; 8];
+        // We could only read 4 bytes in this case
+        client.min_req_size(4).read(&mut bytes)?;
+        assert_eq!(bytes, [b'f', b'g', b'b', 3, b'f', b'g', b'b', 0]);
```

Comment:
So ultimately the client is waiting to issue a second http request here within the call to `read`? I guess that surprises me a little, but it is what it is.

Reply:
It's just one HTTP request, but in this case `read` would allow us to return 4 bytes only. For that, we had to distinguish between `read_all` and `read` In the buffer logic, which isn't worth it, IMO.